pkg/prometheus/operator.go

```diff
@@ -2088,3 +1820,216 @@ func (c *Operator) listMatchingNamespaces(selector labels.Selector) ([]string, e
 	}
 	return ns, nil
 }
+
+type tlsAssetKey struct {
+	from string
+	ns   string
+	name string
+	key  string
+}
+
+func tlsAssetKeyFromSecretSelector(ns string, sel *v1.SecretKeySelector) tlsAssetKey {
+	return tlsAssetKey{
+		from: "configmap",
+		ns:   ns,
+		name: sel.Name,
+		key:  sel.Key,
+	}
+}
+
+func tlsAssetKeyFromConfigMapSelector(ns string, sel *v1.ConfigMapKeySelector) tlsAssetKey {
+	return tlsAssetKey{
+		from: "secret",
+		ns:   ns,
+		name: sel.Name,
+		key:  sel.Key,
+	}
+}
+
+func (k tlsAssetKey) String() string {
+	return fmt.Sprintf("%s_%s_%s_%s", k.from, k.ns, k.name, k.key)
+}
+
+type assetStore struct {
+	cmClient corev1client.ConfigMapsGetter
+	sClient  corev1client.SecretsGetter
+	objStore cache.Store
+
+	tlsAssets         map[tlsAssetKey]TLSAsset
+	bearerTokenAssets map[string]BearerToken
+	basicAuthAssets   map[string]BasicAuthCredentials
+}
+
+func newAssetStore(cmClient corev1client.ConfigMapsGetter, sClient corev1client.SecretsGetter) *assetStore {
+	assetStore := &assetStore{
+		cmClient:          cmClient,
+		sClient:           sClient,
+		tlsAssets:         make(map[tlsAssetKey]TLSAsset),
+		bearerTokenAssets: make(map[string]BearerToken),
+		basicAuthAssets:   make(map[string]BasicAuthCredentials),
+	}
+
+	assetStore.objStore = cache.NewStore(assetStore.keyFunc)
+
+	return assetStore
+}
+
+func (a *assetStore) keyFunc(obj interface{}) (string, error) {
+	switch v := obj.(type) {
+	case *v1.ConfigMap:
+		return fmt.Sprintf("0/%s/%s", v.GetNamespace(), v.GetName()), nil
+	case *v1.Secret:
+		return fmt.Sprintf("1/%s/%s", v.GetNamespace(), v.GetName()), nil
+	}
```

Comment:
client go has `k8s.io/apimachinery/pkg/api/meta.Accessor()` and `k8s.io/apimachinery/pkg/api/meta.TypeAccessor()` functions which let you retrieve the name, namespace, and type, but admittedly this is strictly bound to secrets and configmaps, hence this should be fine.

Reply:
I've looked into this but as you said, the code is so small that it isn't worth the indirection IMO.

Reply:
agreed